torch/csrc/jit/passes/utils/subgraph_matcher.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include <torch/csrc/jit/ir.h>
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace torch {
+namespace jit {
+
+// Usage: instantiate a Graph and use a MatchIterator to iteratively
+// get matches on the underlying graph.  The returned `Match`s can
+// be used as maps from the passed in match graph to the graph the
+// match was run on.
+// 
+//   for (auto m : MatchIterator(match_graph, graph)) {
+//     AT_ASSERT(m[m0.value()] == a.value());
+//   }
+//
+// Optionally, pass in a std::ostream* to get debug information.
+// e.g. `MatchIterator(match_graph, graph, &std::cout)`.
+//
+struct TORCH_API MatchIterator {
+  MatchIterator(
+      const Graph& match_graph,
+      const Graph& graph,
+      std::ostream* log_stream = nullptr)
+      : match_graph_(match_graph),
+        // Arbitrarily choose anchor node
+        anchor_(*(match_graph.nodes().begin())),
+        graph_(graph),
+        finished_(false),
+        log_stream_(log_stream) {}
+
+  // Match is a two typed map that makes it convenient to deal with
+  // the matches by overloading the subscript [] operator to be 
+  // const correct and handle both types (Node*, Value*)
+  struct Match {
```

Comment:
MatchMap for values should be enough as it's trivial to get from `Value` to `Node` and vice versa.

Reply:
its a pretty convenient data structure to have in the implementation, mind if I keep it?

Reply:
I just thought that the main reason of having it was that we had `matchNodes` and `matchValues` at the same time. If it's useful independently on that, then sure we can have it. One reason I slightly prefer unwrapped map from STL to a wrapper is that it's usually safe to assume that a future reader of this code is familiar with standard STL datatypes, but not with custom wrappers (even when they are convenient, the reader have to check their definition to make sure they do what they seem to be doing).